view/readrom.js

```diff
@@ -110,34 +91,55 @@ const prebuilt_json_url_path = "/roms/index.json";
 
 /* Process the index JSON object that contains all the ROMs available */
 function processIndex(index) {
-    const to_option = (entry) => `<option value="${entry.urls}" data-version="${entry.version}" data-hash="${entry.hash}">${entry.name}</option>`;
+    const to_option = (entry,selected=false) => {
+        const attrs = {
+            value: entry.urls,
+            hash: entry.hash,
+        };
+
+        if(selected) {
+            attrs.selected = true;
+        }
+        const attributes = Object.keys(attrs).reduce((acc,key) => {
+            acc += `${key}`;
+            if(!['selected', 'disabled'].includes(key) && attrs[key]) {
+                acc += `="${attrs[key]}" `;
+            }
+            return acc;
+        }, '');
 
-    /* Generate an HTML option out of each entry */
-    const latest  = to_option(index.latest);
-    const nightly = index.nightly.map(to_option);
-    const stable  = index.stable.map(to_option);
+        return `<option ${attributes}>${entry.name}</option>`;
+    };
 
-    const all_options =
-    `<option value="">Choose an image...</option>` +
-    `<optgroup label="--- Latest ---" data-type="latest">` + latest + `</optgroup>` +
-    `<optgroup label="--- Nightly ---"  data-type="nightly">` + nightly.join("") + `</optgroup>` +
-    `<optgroup label="--- Stable ---"  data-type="stable">` + stable.join("") + `</optgroup>`;
+    /* Generate an HTML option out of each entry */
+    const latest  = to_option(index.latest, index.latest.urls == params.r || params.r == 'latest');
+    const nightly = index.nightly.reverse().map((entry, index) => {
+        if((params.r == entry.urls) || (params.r == 'nightly' && index == 0)) {
+            return to_option(entry, true);
+        }
+        return to_option(entry, false);
+    });
+    const stable  = index.stable.map((entry) => to_option(entry, false));
 
+    let all_options = [
+        `<option value="">Choose an image...</option>`,
+    ];
+    if(index.latest) {
+        all_options.push(`<optgroup label="--- Latest ---" data-type="latest">` + latest + `</optgroup>`);
+    }
+    if(index.nightly) {
```

Comment:
An index is unlikely to belong to `nightly`, `latest` and `stable`, so using `else if` seems more appropriate

Reply:
Not sure I understand - “index” here is the object that holds the latest, nightly and stable properties

these checks are just making sure those properties exist before attempting to access them

Reply:
My bad, ignore my comment, I misunderstood it